packages/core/saved-objects/core-saved-objects-migration-server-internal/src/document_migrator/pipelines/downgrade_pipeline.ts

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import Boom from '@hapi/boom';
+import { cloneDeep } from 'lodash';
+import Semver from 'semver';
+import type { SavedObjectUnsanitizedDoc } from '@kbn/core-saved-objects-server';
+import { Transform, TransformType, TypeTransforms } from '../types';
+import type { MigrationPipeline, MigrationPipelineResult } from './types';
+import { applyVersion, assertValidCoreVersion } from './utils';
+
+export class DocumentDowngradePipeline implements MigrationPipeline {
```

Comment:
I spent a *lot* of time trying to figure out if it was better to have the logic for both up and down conversions at the same place or them being dissociated, and at the end, I think this is better:
- The up migration has a lot of legacy logic because of the `Reference` and `Converting` migrations that down migrations don't have to care about
- The logic and sanity check are fully different (e.g `assertUpgrade` in the up pipeline)
- Even the look and version-applying logic are somewhat different

So I went with the 2 distinct logics approach, and I still think that's the correct decision.